docs/exceptions.md

```diff
@@ -0,0 +1,266 @@
+trickery to avoid them. The decision process for this is out of scope of this
+document.
+
+### Regular error
+
+Consider a function like `readFile`. It is entirely possible that the path
+passed to it may point to a file that does not exist, and the programmer can't
+do anything to prevent this. Existence of a file is not a property of our code,
+it's the property of the outside world, and we have to deal with all possible
+scenarios.
+
+Another example is parsing user input. We might expect the user to enter a
+number, but we have to consider the case that the user enters something else.
+
+In cases when the erroneous scenarios are out of our control, we consider
+these to be regular errors.
+
+### Identifying error class
+
+Can the execution of the error codepath be excluded by code review and static
+verification (more precise types?)
+
+* Yes: it is a programmer mistake
+* No: it is a regular error
+
+## Is the code 100% pure or potentially impure?
+
+### Impure code
+
+The code is considered impure when it's an `IO` action, a function that returns
+an `IO` action, or similar. The code can be _potentially_ impure when it is
+written in an abstract monad that can be instantiated to `IO`.
+
+Definitely impure:
+
+```
+readFile :: FilePath -> IO String
+```
+
+Potentially impure:
+
+```
+lookupThrow :: MonadThrow m => Map k v -> k -> m v
+```
+
+### Pure code
+
+Pure code is one that does not use `IO` operations. For example:
+
+```
+lookupMaybe :: Map k v -> k -> Maybe v
+```
+
+Notice that while it is conceivable that we have a `Map` of `IO` actions, and
+instantiate `lookupMaybe` to `Map k (IO ()) -> k -> Maybe (IO ())`, the
+instantiation does not inspect/use these `IO` actions, so it's not considered
+potentially impure.
+
+### Identifying code purity
+
+Is there an instantiation of type parameters that would mean that there are
+`IO` actions inspected/used in the code?
+
+* Yes: it is potentially impure code
+* No: it is 100% pure code
+
+## Error handling practices
+
+### Pure code, programmer mistakes
+
+Do *not*:
+
+* do this often, as if it's okay
+* use `undefined`
+* use non-exhaustive pattern matching
+
+Do:
+
+* try to use types to avoid the need in the first place
+* comment extensively (invariants and precondition, reasoning)
+* use the `error` function (or `impureThrow` with custom exception)
+* use the `HasCallStack` feature
+
+Before writing code that allows programmer mistakes, consult with colleagues how
+it would be possible to prevent them statically. There might be tricks that
+you're not aware of. Always make this trade-off conciously and responsibly.
+
+(On the other hand, if static guarantees require GADTs or type families, perhaps
+it's better to not overcomplicate code. Use your judgement.)
+
+DISCUSSION: Should we create a synonym `bug = impureThrow` in Universum? This
+would make the intention more clear.
+
+### Pure code, regular errors
```

Comment:
Regular error means

> In cases when the erroneous scenarios are out of our control, we consider these to be regular errors.

but in a pure program, nothing is out of our control, so does "Pure code, regular errors" make any sense?

Reply:
Well, technically yes, we do control what our function take as input. But consider these three ways to give a type signature to `head`:

1. `[a] -> a`
2. `NonEmpty a -> a`
3. `[a] -> Maybe a

In case (1) we assume that the caller has full control of the passed argument, and passing `[]` is a programmer error. If the provenance of the input list external (i.e. we've read the list from a file), we (as a caller) need to check whether the list is empty. In case (2) we ensure this invariant on types. But in case (3) we give move the check into the function itself, rather than impose it on the caller. You could say that this is not an "error path", but informally you would normally think about this as a failure (we tried to extract the `head`, but there isn't one).

If we tried to formalize this, I would say in a pure function we can talk about an expected set of codepaths, and we also have checks that ensure that the input satisfies the necessary conditions for this codepath. When the checks fail, we want early from the function — this is what I mean by "regular errors in pure code".

I agree that it's not a strict definition, it depends on what you consider to be an error, but that's how we usually reason about code informally.

Reply:
The lesson here is that `head :: [a] -> a` is a lie :) 

I revise my statement then: in a *total* pure program, error throwing / handling doesn't make sense. There can be programmer mistakes, sure, but the function is total so in these cases the programmer didn't realize there was a mistake.

Reply:
Even in total functions, can't we talk about having a "major" codepath and various "error" codepaths? For instance, when we consider parsing, "parse errors" are called "errors" for a reason, no?

Reply:
You're right it's confusing terminology. It's a parse error, but a *parse error* is not an *error in parsing*. Parsing went just fine: we successfully determined that the input string is not in the language. A better terminology would perhaps be `parse :: Parser a -> Either Unrecognized a`.

Reply:
I tend to use "error" to mean a programmer mistake i.e. a bug. Something that could be avoided but was not, because we're human. So regarding `head :: [a] -> a`, yes, the error would be the caller's error, because he was just supposed to know that the input must not be `[]`.

I tend to use "exception" to mean something that the programmer could not possibly have defended against / ruled out and is only capable of handling / responding to. There's no exception in `head []`, it's a plain error. But reading from a TCP socket that has been reset is an exception. The programmer did nothing wrong, the connection just broke. Failing to handle an exception could be considered an error.

I guess you could also speak of "user error" but that's very different. It doesn't mean any errors in our program. A parse error could be a user error (user gave a malformed config file or something). It can be considered an error to fail to deal with a user error.